File: src/main/java/uk/co/reecedunn/intellij/plugin/xquery/resolve/reference/XQueryEQNamePrefixReference.java
package uk.co.reecedunn.intellij.plugin.xquery.resolve.reference;

import com.intellij.openapi.util.TextRange;
import com.intellij.psi.*;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import uk.co.reecedunn.intellij.plugin.xquery.ast.xquery.XQueryEQName;
import uk.co.reecedunn.intellij.plugin.xquery.psi.XQueryNamespace;
import uk.co.reecedunn.intellij.plugin.xquery.psi.XQueryNamespaceProvider;

public class XQueryEQNamePrefixReference extends PsiReferenceBase<XQueryEQName> {
    private CharSequence mPrefix;

    public XQueryEQNamePrefixReference(XQueryEQName element, TextRange range, CharSequence prefix) {
        super(element, range);
        mPrefix = prefix;
    }

    @Nullable
    @Override
    public PsiElement resolve() {
        PsiElement element = getElement();
        return resolvePrefix(element);
    }

    @NotNull
    @Override
    public Object[] getVariants() {
        return new Object[0];
    }

    private PsiElement resolvePrefix(PsiElement element) {
        while (element != null) {
            if (element instanceof XQueryNamespaceProvider) {
                XQueryNamespace resolved = ((XQueryNamespaceProvider) element).resolveNamespace(mPrefix);
                if (resolved != null) {
                    return resolved.getPrefix();
                }
            }

            PsiElement next = element.getPrevSibling();
            if (next == null) {
                next = element.getParent();
            }
            element = next;
        }
        return null;
    }
}

File: src/main/java/uk/co/reecedunn/intellij/plugin/xquery/resolve/reference/XQueryEQNamePrefixReference.kt
/*
 * Copyright (C) 2016-2017 Reece H. Dunn
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.reecedunn.intellij.plugin.xquery.resolve.reference

import com.intellij.openapi.util.TextRange
import com.intellij.psi.PsiElement
import com.intellij.psi.PsiReferenceBase
import uk.co.reecedunn.intellij.plugin.xpath.ast.xpath.XPathEQName

class XQueryEQNamePrefixReference(element: XPathEQName, range: TextRange) : PsiReferenceBase<XPathEQName>(element, range) {
    override fun resolve(): PsiElement? {
        return element.resolvePrefixNamespace().map { ns -> ns.namespacePrefix as PsiElement }.firstOrNull()
    }

    override fun getVariants(): Array<Any> {
        return arrayOf()
    }
}
